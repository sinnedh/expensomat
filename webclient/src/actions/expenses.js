export const getExpenses = (token) => {
  return {type: 'EXPENSES:LOAD_REQUEST', token}
}

export const createExpense = (token, expense) => {
  return {type: 'EXPENSES:CREATE_REQUEST', token, expense}
}

export const deleteExpense = (token, expense) => {
  return {type: 'EXPENSES:DELETE_REQUEST', token, expense}
}

export const updateExpense = (token, id, changes) => {
  return {type: 'EXPENSES:UPDATE_REQUEST', token, id, changes}
}
