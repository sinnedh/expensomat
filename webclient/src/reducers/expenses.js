import { List, Map } from 'immutable'

const initialState = Map()

const expenses = (state = initialState, action) => {
  switch (action.type) {
    case 'EXPENSES:LOAD_SUCCESS':
      return List(action.items).reduce(
        (result, value) => result.set(value['id'].toString(), Map(value)),
        Map()
      );

    case 'EXPENSES:CREATE_SUCCESS':
      return state.merge({[action.expense.id]: Map({...action.expense})})

    case 'EXPENSES:DELETE_SUCCESS':
      return state.delete(action.id.toString())

    case 'EXPENSES:UPDATE_SUCCESS':
      return state.merge({[action.id]: action.expense})

    case 'EXPENSES:LOAD_REQUEST':
    case 'EXPENSES:LOAD_FAILURE':
    case 'EXPENSES:CREATE_REQUEST':
    case 'EXPENSES:CREATE_FAILURE':
    case 'EXPENSES:DELETE_REQUEST':
    case 'EXPENSES:DELETE_FAILURE':
    case 'EXPENSES:UPDATE_REQUEST':
    case 'EXPENSES:UPDATE_FAILURE':
    default:
      return state;
  }
}

export default expenses;
