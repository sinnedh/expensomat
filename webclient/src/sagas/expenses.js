import { call, put } from 'redux-saga/effects'
import { setErrorNotification, setInfoNotification } from '../actions'
import { incrementLoadingCounter, decrementLoadingCounter } from '../actions'
import * as api from '../api'

export function* createExpense(action) {
  yield put(incrementLoadingCounter())
  try {
    yield call(api.createExpense, action.token, action.expense)
    yield put({
      type: "EXPENSES:CREATE_SUCCESS",
      expense: action.expense,
    })
    yield put(setInfoNotification('Expense created succesfully'))
  } catch (e) {
    yield put({type: "EXPENSES:CREATE_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not create expense ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}

export function* updateExpense(action) {
  yield put(incrementLoadingCounter())
  try {
    const expense = yield call(api.updateExpense, action.token, action.id, action.changes)

    yield put({
      type: "EXPENSES:UPDATE_SUCCESS",
      id: expense.data.id,
      expense: expense.data,
    })
    yield put(setInfoNotification('Expense updated succesfully'))
  } catch (e) {
    yield put({type: "EXPENSES:UPDATE_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not update expense ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}

export function* deleteExpense(action) {
  yield put(incrementLoadingCounter())
  try {
    yield call(api.deleteExpense, action.token, action.id)
    yield put({
      type: "EXPENSES:DELETE_SUCCESS",
      id: action.id,
    })
    yield put(setInfoNotification('Expense deleted succesfully'))
  } catch (e) {
    yield put({type: "EXPENSES:DELETED_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not delete expense ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}

export function* fetchExpenses(action) {
  yield put(incrementLoadingCounter())

  try {
    const expenses = yield call(api.fetchExpenses, action.token)
    yield put({
      type: "EXPENSES:LOAD_SUCCESS",
      items: expenses.data,
    })
  } catch (e) {
    yield put({type: "EXPENSES:LOAD_FAILURE", message: e.message})
    yield put(setErrorNotification(`Could not load expenses ("${e.message}")`))
  } finally {
    yield put(decrementLoadingCounter())
  }
}
