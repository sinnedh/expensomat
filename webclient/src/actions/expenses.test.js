import * as actions from './expenses'

describe('EXPENSES', () => {
  const token = 'ABCD123'
  const expense = {id: 1, name: 'Expense'}

  it('handles getExpenses', () => {
    const action = actions.getExpenses(token)
    expect(action).toMatchSnapshot()
  })

  it('handles createExpense', () => {
    const action = actions.createExpense(token, expense)
    expect(action).toMatchSnapshot()
  })

  it('handles deleteExpense', () => {
    const action = actions.deleteExpense(token, expense.id)
    expect(action).toMatchSnapshot()
  })

  it('handles updateExpense', () => {
    const action = actions.updateExpense(token, 11, {name: 'Expense'})
    expect(action).toMatchSnapshot()
  })
})
