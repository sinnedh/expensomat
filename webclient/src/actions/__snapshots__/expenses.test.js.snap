// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EXPENSES handles createExpense 1`] = `
Object {
  "expense": Object {
    "id": 1,
    "name": "Expense",
  },
  "token": "ABCD123",
  "type": "EXPENSES:CREATE_REQUEST",
}
`;

exports[`EXPENSES handles deleteExpense 1`] = `
Object {
  "id": 1,
  "token": "ABCD123",
  "type": "EXPENSES:DELETE_REQUEST",
}
`;

exports[`EXPENSES handles getExpenses 1`] = `
Object {
  "token": "ABCD123",
  "type": "EXPENSES:LOAD_REQUEST",
}
`;

exports[`EXPENSES handles updateExpense 1`] = `
Object {
  "changes": Object {
    "name": "Expense",
  },
  "id": 11,
  "token": "ABCD123",
  "type": "EXPENSES:UPDATE_REQUEST",
}
`;
